Sources/NIOHTTP1/HTTPPipelineSetup.swift

```diff
@@ -14,23 +14,49 @@
 
 import NIO
 
-/// Configuration required to configure a HTTP pipeline for upgrade.
+/// Configuration required to configure a HTTP client pipeline for upgrade.
+///
+/// See the documentation for `HTTPClientUpgradeHandler` for details on these
+/// properties.
+public typealias NIOHTTPClientUpgradeConfiguration = (upgraders: [NIOHTTPClientProtocolUpgrader], completionHandler: (ChannelHandlerContext) -> Void)
+
+/// Configuration required to configure a HTTP server pipeline for upgrade.
 ///
 /// See the documentation for `HTTPServerUpgradeHandler` for details on these
 /// properties.
-public typealias HTTPUpgradeConfiguration = (upgraders: [HTTPServerProtocolUpgrader], completionHandler: (ChannelHandlerContext) -> Void)
+@available(*, deprecated, renamed: "NIOHTTPServerUpgradeConfiguration")
+public typealias HTTPUpgradeConfiguration = NIOHTTPServerUpgradeConfiguration
+
+public typealias NIOHTTPServerUpgradeConfiguration = (upgraders: [HTTPServerProtocolUpgrader], completionHandler: (ChannelHandlerContext) -> Void)
 
 extension ChannelPipeline {
     /// Configure a `ChannelPipeline` for use as a HTTP client.
     ///
     /// - parameters:
     ///     - position: The position in the `ChannelPipeline` where to add the HTTP client handlers. Defaults to `.last`.
+    ///     - upgrade: Whether to add a `HTTPClientUpgradeHandler` to the pipeline, configured for
+    ///         HTTP upgrade. Defaults to `nil`, which will not add the handler to the pipeline. If
+    ///         provided, should be a tuple of an array of `HTTPClientProtocolUpgrader` and the upgrade
+    ///         completion handler. See the documentation on `HTTPClientUpgradeHandler` for more
+    ///         details.
     /// - returns: An `EventLoopFuture` that will fire when the pipeline is configured.
     public func addHTTPClientHandlers(position: Position = .last,
-                                      leftOverBytesStrategy: RemoveAfterUpgradeStrategy = .dropBytes) -> EventLoopFuture<Void> {
-        return self.addHandlers(HTTPRequestEncoder(),
-                                ByteToMessageHandler(HTTPResponseDecoder(leftOverBytesStrategy: leftOverBytesStrategy)),
-                                position: position)
+                                      leftOverBytesStrategy: RemoveAfterUpgradeStrategy = .dropBytes,
+                                      withClientUpgrade upgrade: NIOHTTPClientUpgradeConfiguration? = nil) -> EventLoopFuture<Void> {
```

Comment:
Unfortunately this change is semver-major, as it changes the fully-qualified name of this function. While it probably doesn't matter, we should aim to preserve the original name as well.

This means we need two functions: one with the old two-argument form that calls into the new one, which has a three-argument form. To avoid ambiguity issues, the three-argument form needs to not have an explicit default argument for `withClientUpgrade`, though the argument type should remain optional.